### ISSUE TITLE
Sync Error due to floating point inconsistency

### ISSUE THREAD

**Issue by LegendOfDragoon** [project64/project64#848]
I was debugging Star Wars Battle for Naboo and got a Sync Error because the interpreter (based on the compiler settings i used) basically does

> _(float *)_FPR_S[m_Opcode.fd] = (_(float *)_FPR_S[m_Opcode.ft] + *(float *)_FPR_S[m_Opcode.fs]);

instead of

> _(float *)_FPR_S[m_Opcode.fd] = (_(float *)_FPR_S[m_Opcode.fs] + *(float *)_FPR_S[m_Opcode.ft]);

for COP1_S_ADD. This difference can cause a change in the results. For instance, when f24 = 0xFFDF9358 and f20 = 0xFFB78B9C, the recompiler's result for f24 = f24 + f20 was 0xFFDF9358 , while the interpreter's result was 0xFFF78B9C. Hopefully there's some way to force the compiler to do the operations in a specific order.

**Comment by LegendOfDragoon** [comment]
Unfortunately, I don't think there is a compiler setting to ensure that it does operations in the specific order.

**Comment by cxd4** [comment]
Floating-point code needs to be rewritten for portability in the interpreter.

Even using SSE intrinsics is way more cautious at avoiding these sorts of FP round-off issues than hacking in undefined behavior like this and type-casting *((float *)(void *)FPR[x]).

**Comment by LegendOfDragoon** [comment]
Looks like disabling optimizations made the interpreter do it in the right order. Also it worked when i replaced m_Opcode.fs and m_Opcode.ft with an alternative algorithm (just manually shifting + ANDing), but I would not rely on doing that, unless there's some known reason behind why the compiler would treat bitfields differently.

How would you write portable floating point code? Would you use unions?

**Comment by cxd4** [comment]
Unions are not really portable either because reading from any member which was not the last one previously written to is considered undefined behavior.

So for example if you have this:
```c
typedef union {
    uint8_t bytes[4];
    uint16_t halfwords[2];
    uint32_t word;
} MIPS_GPR;
```
... and you do ...
```c
GPR[16].word = 0x1234ABCD;
```
... then you cannot really guarantee that GPR[16].halfwords[0] (or even halfwords[1]) is 0x1234 or 0xABCD.  You can only hope that it is for most platforms, which usually like for x86 it is.

In terms of portability, unions make no guarantees other than alignment I think.

With a little concentration on the floating-point precision control from `<float.h>` you should be able to make the FP math more well-defined and portable.  It should be capable of optimizing to SSE2 without the need for intrinsics with most simple things.

**Comment by cxd4** [comment]
Another thing you could do is invent your own floating-point type, to be bit-perfect with the MIPS floating-point specifications.  Instead of using `float`s and `double`s you could create a `struct` and only ever touch the integer mantissa/fraction bitfield within the struct to guarantee avoidance of things like implementation-defined treatment of precision round-off etc.

**Comment by LegendOfDragoon** [comment]
This is a serious problem :disappointed: . I spent about an hour debugging Indiana Jones, just to find out the sync error is caused by obscure differences in floating point code. I now see why zilmar initially used inline assembly. I bet the recent sync error I discovered in Twisted Edge (which weren't present in earlier builds of 2.2) is related to these floating point changes.

Roundf is generating a different result than fistp does. In CVT.S.W, the result was different when the floating point register = 0x4A1B8BB2.

**Comment by project64** [comment]
I made a note I need to test converting float -2.5, -1.5, -0.5, 0.5, 1.5, 2.5 rounding to nearest produces the same expected results. Cause this was a place where it is not documented well how rounding works ..  Once I get the android version out and probably needs a bit more work once released, but after that I would like to look at creating test roms for the real n64 and just see how well pj64 does.

**Comment by LegendOfDragoon** [comment]
Ok, well the first issue I pointed out has to do with handling NANs. Probably neither the recompiler or interpreter handle it correctly. It'll probably require some work to figure out how the N64 handles this.

For this issue, I just would like the recompiler and interpreter to be consistent so that I don't end up spending hours debugging sync errors that are not significant.

As for this new problem I encountered, I think the code may need to be reevaluated. https://github.com/project64/project64/blob/master/Source/Project64-core/N64System/Interpreter/InterpreterOps.cpp#L2388 . I'm thinking maybe you can just remove the roundf since it's already doing a conversion from float to int.

**Comment by project64** [comment]
the asm would be consistent with the recompiler since it is doing the same thing which is good and how it was. But I had to use the roundf so it could be compiled on arm for android, so I can not go back to using asm in the interpreter.

**Comment by cxd4** [comment]
> I'm thinking maybe you can just remove the roundf since it's already doing a conversion from float to int.

Which isn't the same thing as rounding to the nearest integer.

Type-conversion is nastier than functions that way.  `(int)((float)x)` is only well-defined if x is something like 31.0, in which case the result is 31.  In which case the fact remains that this still isn't rounding.  With a custom function, a clean rounded integer can be returned without the need for typecasts that break x86 compilation for x86 fanboys.

**Comment by LegendOfDragoon** [comment]
I guess for now I'll have to just tweak my local copy before the next time I decide to track down some sync errors. I just tried tested Rogue Squadron and it appears to have the same issue as Indiana Jones.

>Which isn't the same thing as rounding to the nearest integer.
Type-conversion is nastier than functions that way.

I see :disappointed: .

**Comment by AmbientMalice** [comment]
I notice the CP1ROUND test from https://github.com/PeterLemon/N64 crashes the emulator, while pretty much every other CP1 test succeeds.

**Comment by JunielKatarn** [comment]
> I notice the CP1ROUND test from https://github.com/PeterLemon/N64 crashes the emulator

@AmbientMalice What is that repo about?

**Comment by LegendOfDragoon** [comment]
>I notice the CP1ROUND test from https://github.com/PeterLemon/N64 crashes the emulator, while pretty much every other CP1 test succeeds.

The recompiler seems fine, but the interpreter actually doesn't pass certain tests. I tried Ceil and it failed some of them. So I guess there really are 2 issues.

**Comment by bryanperris** [comment]
I don't know if this helpful but I have some information on this.  First of, the FPU tests made by PeterLemon all use the nearest rounding mode which is the default.  They all have passed on my C# written interpreter using the standard functions for Round, Ceil, Truncate and Floor, since MIPS and Intel use the nearly the same IEEE 754 standard for floating point arithmetic. I know that PJ64 calls upon fp_control to switch the FPU round mode before executing a FPU operation, and I don't think that is necessary, and some of the modes can cause regression in performance too, correct me if I am wrong.

Reading through the FPU documentation for the MIPS, all main FPU ALU operations first compute the value with the highest precision available then the rounding is applied just before writeback to the FPR register.  So my solution is calling my own round function during each FPU writeback.

public static float Round(float value, MipsRoundingMode mode)
{
	switch (mode)
	{
		default: return value; // Nearest - Default 
		case MipsRoundingMode.Postive: return (float)Math.Ceiling(value);
		case MipsRoundingMode.Negative: return (float)Math.Floor(value);
		case MipsRoundingMode.Zero: return (float)Math.Truncate(value);
	}
}

public static double Round(double value, MipsRoundingMode mode)
{
	switch (mode)
	{
		default: return value; // Nearest - Default 
		case MipsRoundingMode.Postive: return Math.Ceiling(value);
		case MipsRoundingMode.Negative: return Math.Floor(value);
		case MipsRoundingMode.Zero: return Math.Truncate(value);
	}
}

As for the case of float32 -> float64 and vice versa, I also apply the rounding mode.  For converting unsigned integers to floating point, and vice versa, I let the native CPU handle this internal through normal type casting, I do not use any round routine explicitly.  However when I do cast a floating point value to unsigned integers, I make sure to cast it into the signed integer type before the unsigned cast.

HTH

**Comment by LegendOfDragoon** [comment]
When I first opened this issue, there were 2 issues related to floating point that caused sync errors. One was the rounding issue (which has since been fixed), and the other has to do with order of operations. The current problem is that the compiler changes the order of operation, which changes the calculation, if the numbers are NaNs. 

>all main FPU ALU operations first compute the value with the highest precision available then the rounding is applied just before writeback to the FPR register.

Wouldn't that mean it always rounds to whatever precision is currently being used? If I understand correctly, then the rounding method currently used is necessary.

**Comment by bryanperris** [comment]
Based on this table: https://imgur.com/a/6BhGAF1
There are rounding stages that take place at certain places on the FPU pipeline, so who knows if the native CPU has the same layout of these stages.  Also about controlfp:  "On the ARM and x64 architectures, changing the infinity mode or the floating-point precision is not supported"

What I did in my interpreter was execute the native FPU operation then round the result, but of course without test roms using the different rounding modes, I can't say this is a 100% guaranteed solution either.  I have ordered a N64 Drive and waiting for its arrival so that  I can mess around with the real hardware, and I also put in a request Peter Lemon.
